util: Allow `Assert` (et al.) in contexts without __func__

maflcko opened the pull request:
Without this, compile warnings could be hit about `__func__` being only valid inside functions.

```
warning: predefined identifier is only valid inside function [-Wpredefined-identifier-outside-function] note: expanded from macro Assert
  115 | #define Assert(val) inline_assertion_check<true>(val, __FILE__, __LINE__, __func__, #val)
      |                                                                           ^
```

Ref https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2486258473


This also introduces a slight behaviour change, because `std::source_location::function_name` usually includes the entire function signature instead of just the name.

Commit:
util: Allow Assert() in contexts without __func__

Without this, compile warnings could be hit about __func__ being only
valid inside functions.

warning: predefined identifier is only valid inside function [-Wpredefined-identifier-outside-function]
note: expanded from macro Assert
  115 | #define Assert(val) inline_assertion_check<true>(val, __FILE__, __LINE__, __func__, #val)
      |                                                                           ^

Ref https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2486258473

Commit:
refactor: Use const reference to std::source_location

Performance likely does not matter here, but from a perspective of
code-readablilty, a const reference should be preferred for read-only
access.

So use it here.

This requires to set -Wno-error=dangling-reference for GCC 13.1
compilations, but this false-positive is fixed in later GCC versions.

See also https://godbolt.org/z/fjc6be65M

stickies-v reviewed (approved):
ACK fae1d99651e29341e486a10e6340335c71a2144e

Note that this also introduces behaviour change by including the entire function signature instead of just the name. Might be worth mentioning in OP.

hodlinator commented on ci/test/00_setup_env_win64.sh:
remark: Disabling dangling reference warnings globally is not ideal, but chances that it will not be reverted as the container is upgraded seem slim.

maflcko replied:
Correct, it is not ideal. Though, it is only disabled for this specific task and only temporary. Also, the GCC check is mostly or fully redundant anyway with the clang check (which works better due to lifetime annotations), as well as all the runtime sanitizers.

hodlinator replied:
Tested running...
```shell
₿ env -i HOME="$HOME" PATH="$PATH" USER="$USER" bash -c 'MAKEJOBS="-j$(nproc)" FILE_ENV="./ci/test/00_setup_env_win64.sh" ./ci/test_run_all.sh'
```
...without the modification to the env shell scripts. Got the expected error:
```
C++ compiler .......................... GNU 13.0.0, /bin/x86_64-w64-mingw32-g++-posix
...
/ci_container_base/src/test/blockmanager_tests.cpp: In member function ‘void blockmanager_tests::blockmanager_scan_unlink_already_pruned_files::test_method()’:
/ci_container_base/src/test/blockmanager_tests.cpp:63:17: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
   63 |     const auto& chainman = Assert(m_node.chainman);
      |                 ^~~~~~~~
In file included from /ci_container_base/src/streams.h:13,
                 from /ci_container_base/src/dbwrapper.h:11,
                 from /ci_container_base/src/node/blockstorage.h:10,
                 from /ci_container_base/src/test/blockmanager_tests.cpp:8:
/ci_container_base/src/util/check.h:116:49: note: the temporary was destroyed at the end of the full expression ‘inline_assertion_check<true, std::unique_ptr<ChainstateManager>&>(((blockmanager_tests::blockmanager_scan_unlink_already_pruned_files*)this)->blockmanager_tests::blockmanager_scan_unlink_already_pruned_files::<anonymous>.TestChain100Setup::<anonymous>.TestingSetup::<anonymous>.ChainTestingSetup::<anonymous>.BasicTestingSetup::m_node.node::NodeContext::chainman, std::source_location{(& *.Lsrc_loc27)}, std::basic_string_view<char>(((const char*)"m_node.chainman")))’
  116 | #define Assert(val) inline_assertion_check<true>(val, std::source_location::current(), #val)
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ci_container_base/src/test/blockmanager_tests.cpp:63:28: note: in expansion of macro ‘Assert’
   63 |     const auto& chainman = Assert(m_node.chainman);
      |                            ^~~~~~
cc1plus: all warnings being treated as errors
gmake[2]: Leaving directory '/ci_container_base/ci/scratch/build-x86_64-w64-mingw32'
gmake[2]: *** [src/test/CMakeFiles/test_bitcoin.dir/build.make:382: src/test/CMakeFiles/test_bitcoin.dir/blockmanager_tests.cpp.obj] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1810: src/test/CMakeFiles/test_bitcoin.dir/all] Error 2
gmake[1]: Leaving directory '/ci_container_base/ci/scratch/build-x86_64-w64-mingw32'
gmake: *** [Makefile:146: all] Error 2
```
Makes sense as this PR is doing, changing the warning to no longer be an error, but still output it as a reminder, as is already done for `-Wno-error=maybe-uninitialized`.

l0rinc replied:
> Removing const over a bunch of places

k, I had the impression this was the only one 👍

hodlinator reviewed (approved):
ACK fae1d99651e29341e486a10e6340335c71a2144e

#### Tested inserting `CHECK_NONFATAL` to see how output differed

Was concerned that `std::source_location::current()` inside of macros might behave slightly differently than `__func__` etc. But only expected differences where observed.

```diff
--- a/src/script/miniscript.cpp
+++ b/src/script/miniscript.cpp
@@ -17,6 +17,7 @@ namespace miniscript {
 namespace internal {
 
 Type SanitizeType(Type e) {
+    CHECK_NONFATAL(false);
     int num_types = (e << "K"_mst) + (e << "V"_mst) + (e << "B"_mst) + (e << "W"_mst);
     if (num_types == 0) return ""_mst; // No valid type, don't care about the rest
     CHECK_NONFATAL(num_types == 1); // K, V, B, W all conflict with each other
```

##### Before

```shell
₿ build/bin/bitcoin-cli -regtest deriveaddresses "wsh(1)#mrg7xj7p"
error code: -1
error message:
Internal bug detected: false
../src/script/miniscript.cpp:20 (SanitizeType)
...
```

##### After

```shell
₿ build/bin/bitcoin-cli -regtest deriveaddresses "wsh(1)#mrg7xj7p"
error code: -1
error message:
Internal bug detected: false
../src/script/miniscript.cpp:20 (miniscript::Type miniscript::internal::SanitizeType(miniscript::Type))
...
```

l0rinc commented on src/util/check.h:
```diff
@@ -124,14 +125,14 @@ constexpr T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] con
 /**
  * NONFATAL_UNREACHABLE() is a macro that is used to mark unreachable code. It throws a NonFatalCheckError.
  */
 #define NONFATAL_UNREACHABLE()                                        \
```
nit: can we align the formatting - either in [column](https://github.com/bitcoin/bitcoin/blob/28d679bad9fda3f180ab0f7d34353e1fa9294d68/src/logging.h#L384-L390) or at the [end of each line](https://github.com/bitcoin/bitcoin/blob/af78d36512999be0ea59715ac5de8cb8d0b3d004/src/util/check.h#L111-L112), but this in-between looks weird

l0rinc reviewed (changes requested):
Concept ACK

It seems to me we should do a deeper cleanup now that `__func__` isn't used, apply it to a few more cases and not use a global suppression but adjust the call-sites instead. 

nit: it seems to me it's not just `Assert` that's affected (as the PR title claims), but `Assume` as well.

Commit:
doc: Remove unused bugprone-lambda-function-name suppression

Now that the __func__ is no longer used, the
NOLINTBEGIN(bugprone-lambda-function-name) can be removed.

Also, re-format the NONFATAL_UNREACHABLE macro, while touching the
adjacent line.

maflcko commented on src/util/check.h:
```diff
@@ -109,10 +110,10 @@ constexpr T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] con
  * caller, which can then report the issue to the developers.
  */
 #define CHECK_NONFATAL(condition) \
```
Thx, done for all in a new, unrelated, commit.

Commit:
refactor: Use STR_INTERNAL_BUG macro where possible

This ensures a uniform bug template and allows to drop includes and
logic at the call sites.

l0rinc reviewed (approved):
Code review ACK fad6efd3bef1d123f806d492f019e29530b03a5e

The only nit I still have is the global suppression, but we can also just tackle that in a follow-up if needed.

stickies-v reviewed (approved):
ACK fad6efd3bef1d123f806d492f019e29530b03a5e

hodlinator reviewed (approved):
re-ACK fad6efd3bef1d123f806d492f019e29530b03a5e

2 added commits since previous review (https://github.com/bitcoin/bitcoin/pull/33785#pullrequestreview-3427192513):
* Remove no longer needed disabling of linter according to CI. 👍 (Re-touching `NONFATAL_UNREACHABLE` but no big deal).
* De-duplicate code by using `STR_INTERNAL_BUG()`.

danielabrozzoni commented:
post-merge concept ACK, thanks for working on this! :)
